helix-core/src/main/java/org/apache/helix/manager/zk/ZkBucketDataAccessor.java

```diff
@@ -0,0 +1,178 @@
+package org.apache.helix.manager.zk;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.I0Itec.zkclient.serialize.ZkSerializer;
+import org.apache.helix.AccessOption;
+import org.apache.helix.BaseDataAccessor;
+import org.apache.helix.BucketDataAccessor;
+import org.apache.helix.HelixException;
+import org.apache.helix.HelixProperty;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.manager.zk.client.DedicatedZkClientFactory;
+import org.apache.helix.manager.zk.client.HelixZkClient;
+import org.apache.helix.util.GZipCompressionUtil;
+import org.codehaus.jackson.map.ObjectMapper;
+
+public class ZkBucketDataAccessor implements BucketDataAccessor {
+
+  private final static ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+  private static final String BUCKET_SIZE_KEY = "BUCKET_SIZE";
+  private static final String DATA_SIZE_KEY = "DATA_SIZE";
+
+  // 100 KB for default bucket size
+  private static final int DEFAULT_BUCKET_SIZE = 100 * 1024;
+  private int _bucketSize = DEFAULT_BUCKET_SIZE;
+  private ZkSerializer _zkSerializer;
+  private BaseDataAccessor _zkBaseDataAccessor;
+
+  public ZkBucketDataAccessor(String zkAddr) {
```

Comment:
I don't like the idea of doing more than necessary initialization works at the constructor level. It should be whoever uses the class' concerns. 

A simple setter like
```
public ZkBucketDataAccessor(ZkSerializer zkSerializer, BaseDataAccessor baseDataAccessor) {
   _zkSerializer = zkSerializer;
   _zkBaseDataAccessor = zkBaseDataAccessor;
}
```

Reply:
I had something different in mind. My idea was make this ZkBucketDataAccessor's implementation details (such as how we serialize or what data accessor we use) hidden from the caller. This way we simplify a lot of code on the caller side and ZkBucketDataAccessor is free to use whichever method it wants to persist the data in ZK.

To put it another way, the rebalancer shouldn't have to worry about how the data is serialized and how we are accessing ZK.